File: 백준/Gold/13905. 세부/세부.py
import heapq


def find(x) -> int:
    if x != parent[x]:
        parent[x] = find(parent[x])
    return parent[x]


def union(x, y):
    root_x = find(x)
    root_y = find(y)

    if root_x < root_y:
        parent[root_y] = root_x
    elif root_x > root_y:
        parent[root_x] = root_y


n, m = map(int, input().split())
s, e = map(int, input().split())

edges = []
for _ in range(m):
    v1, v2, w = map(int, input().split())
    heapq.heappush(edges, (-w, v1, v2))

parent = [i for i in range(n + 1)]

answer = 0
while edges:
    edge = heapq.heappop(edges)
    w = -edge[0]
    v1 = edge[1]
    v2 = edge[2]

    if find(v1) != find(v2):
        union(v1, v2)

        if find(s) == find(e):
            answer = w
            break

print(answer)

File: 백준/Gold/13905. 세부/세부.java
import java.io.*;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int[] parent;
    
    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int s = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());
        PriorityQueue<Edge> edges = new PriorityQueue<>((e1, e2) -> e2.weight - e1.weight);
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int vertex1 = Integer.parseInt(st.nextToken());
            int vertex2 = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());
            edges.add(new Edge(vertex1, vertex2, weight));
        }
        parent = new int[n + 1];
        for (int i = 1; i < n + 1; i++) {
            parent[i] = i;
        }
        
        int minWeight = 0;
        while (!edges.isEmpty()) {
            Edge edge = edges.poll();
            if (find(edge.vertex1) != find(edge.vertex2)) {
                union(edge.vertex1, edge.vertex2);
                
                if (find(s) == find(e)) {
                    minWeight = edge.weight;
                    break;
                }
            }
        }
        bw.write(minWeight + "\n");
        bw.flush();
        bw.close();
    }
    
    static int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
    
    static void union(int x, int y) {
        int xParent = find(x);
        int yParent = find(y);
        
        if (xParent < yParent) {
            parent[yParent] = xParent;
        } else if (xParent > yParent) {
            parent[xParent] = yParent;
        }
    }
    
    static class Edge {
        
        int vertex1;
        int vertex2;
        int weight;
        
        public Edge(int vertex1, int vertex2, int weight) {
            this.vertex1 = vertex1;
            this.vertex2 = vertex2;
            this.weight = weight;
        }
    }
}